src/middlewared/middlewared/plugins/filesystem_/acl_freebsd.py

```diff
@@ -0,0 +1,751 @@
+            })
+        acl.extend((ACLDefault[acl_type].value)['acl'])
+
+        return acl
+
+    def _is_inheritable(self, flags):
+        """
+        Takes ACE flags and return True if any inheritance bits are set.
+        """
+        inheritance_flags = ['FILE_INHERIT', 'DIRECTORY_INHERIT', 'NO_PROPAGATE_INHERIT', 'INHERIT_ONLY']
+        for i in inheritance_flags:
+            if flags.get(i):
+                return True
+
+        return False
+
+    @private
+    def canonicalize_acl_order(self, acl):
+        """
+        Convert flags to advanced, then separate the ACL into two lists. One for ACEs that have been inherited,
+        one for aces that have not been inherited. Non-inherited ACEs take precedence
+        and so they are placed first in finalized combined list. Within each list, the
+        ACEs are orderd according to the following:
+
+        1) Deny ACEs that apply to the object itself (NOINHERIT)
+
+        2) Deny ACEs that apply to a subobject of the object (INHERIT)
+
+        3) Allow ACEs that apply to the object itself (NOINHERIT)
+
+        4) Allow ACEs that apply to a subobject of the object (INHERIT)
+
+        See http://docs.microsoft.com/en-us/windows/desktop/secauthz/order-of-aces-in-a-dacl
+
+        The "INHERITED" bit is stripped in filesystem.getacl when generating a BASIC flag type.
+        It is best practice to use a non-simplified ACL for canonicalization.
+        """
+        inherited_aces = []
+        final_acl = []
+        non_inherited_aces = []
+        for entry in acl:
+            entry['flags'] = self.__convert_to_adv_flagset(entry['flags']['BASIC']) if 'BASIC' in entry['flags'] else entry['flags']
+            if entry['flags'].get('INHERITED'):
+                inherited_aces.append(entry)
+            else:
+                non_inherited_aces.append(entry)
+
+        if inherited_aces:
+            inherited_aces = sorted(
+                inherited_aces,
+                key=lambda x: (x['type'] == 'ALLOW', self._is_inheritable(x['flags'])),
+            )
+        if non_inherited_aces:
+            non_inherited_aces = sorted(
+                non_inherited_aces,
+                key=lambda x: (x['type'] == 'ALLOW', self._is_inheritable(x['flags'])),
+            )
+        final_acl = non_inherited_aces + inherited_aces
+        return final_acl
+
+    @private
+    def getacl_posix1e(self, path, simplified):
+        st = os.stat(path)
+        ret = {
+            'uid': st.st_uid,
+            'gid': st.st_gid,
+            'acl': [],
+            'flags': {
+                'setuid': bool(st.st_mode & pystat.S_ISUID),
+                'setgid': bool(st.st_mode & pystat.S_ISGID),
+                'sticky': bool(st.st_mode & pystat.S_ISVTX),
+            },
+            'acltype': ACLType.POSIX1E.name
+        }
+
+        ret['uid'] = st.st_uid
+        ret['gid'] = st.st_gid
+
+        gfacl = subprocess.run(['getfacl', '-c' if osc.IS_LINUX else '-q', '-n', path],
```

Comment:
Since this only runs on freebsd we may wabt to get rid of this if?

Reply:
> Wonderf if we should have a common class here with the methods definitions so we dont repeat ourselves and also so the API doesnt diverge?
> 
> Thoughts? @themylogin

I think that's a good idea. I'll create a base with the high-level function definitions (like getacl, setacl, etc).